src/Rules/Statements/EchoRule.php

```diff
@@ -0,0 +1,46 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Rules\Statements;
+
+use PhpParser\Node;
+use PHPStan\Analyser\Scope;
+use PHPStan\Rules\Rule;
+use PHPStan\Type\ErrorType;
+use PHPStan\Type\VerbosityLevel;
+
+class EchoRule implements Rule
+{
+
+	public function getNodeType(): string
+	{
+		return Node\Stmt\Echo_::class;
+	}
+
+	/**
+	 * @param Node $node
+	 * @param Scope $scope
+	 * @return string[]
+	 */
+	public function processNode(Node $node, Scope $scope): array
+	{
+		$messages = [];
+
+		foreach ($node->exprs as $key => $expr) {
+			$type = $scope->getType($expr);
+
+			if (!$type->toString() instanceof ErrorType) {
```

Comment:
This is the right approach! You should also skip `Concat`, `Encapsed` and `Cast\String_` in `$expr` because they are covered by different rules.

Reply:
Also, the rule should respect levels. I recommend you to make yourself familiar with `RuleLevelHelper::findTypeToCheck()`. What you want to achieve is to report always wrong types (like `array`) on level 2 and partly wrong types (like `array|string`) on level 6. If you implement RuleLevelHelper call correctly this will be handled automatically.

You can test this by going to https://github.com/phpstan/phpstan/tree/master/tests/PHPStan/Levels and adding another `.php` and adding it as a "topic" to LevelsIntegrationTest.php. When you run the test suite, the `json` report files are generated for you. On subsequent runs, existing files are compared with the actual result - if they are the same, the test passes, if they are different, they get regenerated and the test fails.